terraform/modules/upload-service/batch_watcher.tf

```diff
@@ -0,0 +1,125 @@
+resource "aws_iam_role" "batch_watcher_lambda" {
+  name = "dcp-upload-batch-watcher-daemon-${var.deployment_stage}"
+  assume_role_policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+POLICY
+}
+
+resource "aws_iam_role_policy" "batch_watcher_lambda" {
+  name = "dcp-upload-batch-watcher-daemon-${var.deployment_stage}"
+  role = "${aws_iam_role.batch_watcher_lambda.name}"
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "LambdaLogging",
+      "Action": [
+        "logs:CreateLogGroup",
+        "logs:CreateLogStream",
+        "logs:PutLogEvents",
+        "logs:DescribeLogStreams"
+      ],
+      "Resource": [
+        "arn:aws:logs:*:*:*"
+      ],
+      "Effect": "Allow"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "secretsmanager:DescribeSecret",
+        "secretsmanager:GetSecretValue"
+      ],
+      "Resource": [
+        "arn:aws:secretsmanager:${local.aws_region}:${local.account_id}:secret:dcp/upload/${var.deployment_stage}/*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "batch:Describe*"
+      ],
+      "Resource": [
+        "*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:DescribeInstances",
+        "ec2:TerminateInstances"
+      ],
+      "Resource": [
+        "*"
```

Comment:
Why is this permission (and the one above) so broad?

Reply:
Unfortunate I don't see a way around this. Batch policies cannot be tied to a specific resource. https://github.com/HumanCellAtlas/upload-service/blob/batch-overlord/upload/lambdas/batch_watcher/batch_watcher.py#L50 and https://github.com/HumanCellAtlas/upload-service/blob/batch-overlord/upload/lambdas/batch_watcher/batch_watcher.py#L83 require these broad permissions as they are querying all currently runnning instances. I don't see a way to set a narrow permission that only includes upload spot instances as targets

Reply:
I could restrict it to a region at the least. I'll give it some more thought if this could be restricted to just upload spot instances

Reply:
It doesn't seem that we can restrict these resource paths based on the feature

Reply:
👍